### PR TITLE
Adding build and release pipelines

### PR DESCRIPTION
**Pull request by mimartin12**
Adds the following.

* Build pipeline that will run CLOC, creates an artifact from source, and builds docker images.
* Release pipeline that tags docker then creates docker and github releases.
* Adds in `<Version>` tag to `KeyConnector.csproj` and sets it to `1.0.0` as a base for incrementing release versions.

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+      - name: Setup service name
+        id: setup
+        run: |
+          SERVICE_NAME="key-connector"
+          echo "SERVICE_NAME: $SERVICE_NAME"
+          echo "::set-output name=service_name::$SERVICE_NAME"
+
+      - name: Get build artifact
+        uses: actions/download-artifact@3be87be14a055c47b01d3bd88f8fe02320a9bb60 # v2.0.10
```
Comment (vgrassia) [inline .github/workflows/build.yml:87]:
```suggestion
        uses: actions/download-artifact@3be87be14a055c47b01d3bd88f8fe02320a9bb60  # v2.0.10
```

Comment (vgrassia) [reply .github/workflows/build.yml:87]: To make the linter happy.

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+      - name: Retrieve secrets
+        id: retrieve-secrets
+        uses: Azure/get-keyvault-secrets@80ccd3fafe5662407cc2e55f202ee34bfff8c403
+        if: failure()
+        with:
+          keyvault: "bitwarden-prod-kv"
+          secrets: "devops-alerts-slack-webhook-url"
+
+      - name: Notify Slack on failure
+        uses: act10ns/slack@e4e71685b9b239384b0f676a63c32367f59c2522 # v1.2.2
```
Comment (vgrassia) [inline .github/workflows/release.yml:162]:
```suggestion
        uses: act10ns/slack@e4e71685b9b239384b0f676a63c32367f59c2522  # v1.2.2
```

Comment (vgrassia) [reply .github/workflows/release.yml:162]: To make the linter happy.

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+      - name: Install cloc
+        run: sudo apt update && sudo apt install cloc -y
+
+      - name: Print lines of code
+        run: |
+          cloc --include-lang \
+          C#,SQL,Razor,"Bourne Shell",PowerShell,HTML,CSS,Sass,JavaScript,TypeScript \
+          --vcs git
+
```
Comment (vgrassia) [inline .github/workflows/build.yml:22]:
```suggestion


```

Comment (vgrassia) [reply .github/workflows/build.yml:22]: Two blank lines between jobs.  The linter didn't catch this, so I'll have to check why.

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+          pwd
+          ls -atlh ../../../
+
+      - name: Upload service artifact
+        uses: actions/upload-artifact@82c141cc518b40d92cc801eee768e7aafc9c2fa2
+        with:
+          name: KeyConnector.zip
+          path: ./src/KeyConnector/KeyConnector.zip
+          if-no-files-found: error
+
```
Comment (vgrassia) [inline .github/workflows/build.yml:62]:
```suggestion


```

Comment (vgrassia) [reply .github/workflows/build.yml:62]: Two blank lines between jobs.  The linter didn't catch this, so I'll have to check why.

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        with:
+          azure-creds: ${{ secrets.AZURE_PROD_KV_CREDENTIALS }}
+          azure-keyvault-name: "bitwarden-prod-kv"
+
+      - name: Setup service name
+        id: setup
+        run: |
+          SERVICE_NAME="key-connector"
+          echo "SERVICE_NAME: $SERVICE_NAME"
+          echo "::set-output name=service_name::$SERVICE_NAME"
```
Comment (vgrassia) [inline .github/workflows/build.yml:84]: Delete and replace with job `env` key with `_SERVICE_NAME` subkey.  I'll put an example above.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        uses: actions/upload-artifact@82c141cc518b40d92cc801eee768e7aafc9c2fa2
+        with:
+          name: KeyConnector.zip
+          path: ./src/KeyConnector/KeyConnector.zip
+          if-no-files-found: error
+
+  build-docker:
+    name: Build Docker images
+    runs-on: ubuntu-20.04
+    needs: build-artifacts
```
Comment (vgrassia) [inline .github/workflows/build.yml:66]:
```suggestion
    needs: build-artifacts
    env:
      _SERVICE_NAME: key-connector
```

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+      - name: Setup build artifact
+        run: |
+          mkdir -p ./src/KeyConnector/obj/build-output/publish
+          unzip KeyConnector.zip \
+            -d ./src/KeyConnector/obj/build-output/publish
+
+      - name: Build Docker images
+        run: |
+          docker build -t ${{ steps.setup.outputs.service_name }} \
```
Comment (vgrassia) [inline .github/workflows/build.yml:99]:
```suggestion
          docker build -t ${{ env._SERVICE_NAME }} \
```

File: .github/workflows/build.yml [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+            ./src/KeyConnector
+
+      - name: Tag and Push RC to Docker Hub
+        if: (github.ref == 'refs/heads/rc')
+        run: |
+          docker tag ${{ steps.setup.outputs.service_name }} \
+            bitwarden/${{ steps.setup.outputs.service_name }}:rc
+          docker push bitwarden/${{ steps.setup.outputs.service_name }}:rc
```
Comment (vgrassia) [inline .github/workflows/build.yml:107]:
```suggestion
          docker tag ${{ env._SERVICE_NAME }} \
            bitwarden/${{ env._SERVICE_NAME }}:rc
          docker push bitwarden/${{ env._SERVICE_NAME }}:rc
```

File: .github/workflows/build.yml [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+
+      - name: Tag and Push Hotfix to Docker Hub
+        if: (github.ref == 'refs/heads/hotfix')
+        run: |
+          docker tag ${{ steps.setup.outputs.service_name }} \
+            bitwarden/${{ steps.setup.outputs.service_name }}:hotfix
+          docker push bitwarden/${{ steps.setup.outputs.service_name }}:hotfix
```
Comment (vgrassia) [inline .github/workflows/build.yml:114]:
```suggestion
          docker tag ${{ env._SERVICE_NAME }} \
            bitwarden/${{ env._SERVICE_NAME }}:hotfix
          docker push bitwarden/${{ env._SERVICE_NAME }}:hotfix
```

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+      - name: Tag and Push Dev to Docker Hub
+        if: (github.ref == 'refs/heads/master')
+        run: |
+          docker tag ${{ steps.setup.outputs.service_name }} \
+            bitwarden/${{ steps.setup.outputs.service_name }}:dev
+          docker push bitwarden/${{ steps.setup.outputs.service_name }}:dev
```
Comment (vgrassia) [inline .github/workflows/build.yml:121]:
```suggestion
          docker tag ${{ env._SERVICE_NAME }} \
            bitwarden/${{ env._SERVICE_NAME }}:dev
          docker push bitwarden/${{ env._SERVICE_NAME }}:dev
```

File: .github/workflows/build.yml [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    name: Build Docker images
+    runs-on: ubuntu-20.04
+    needs: build-artifacts
+    steps:
+      - name: Checkout repo
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579
+
+      - name: Setup DCT
+        if: github.ref == 'refs/heads/master' || github.ref == 'refs/heads/rc' || github.ref == 'refs/heads/hotfix'
+        id: setup-dct
```
Comment (vgrassia) [inline .github/workflows/build.yml:73]:
```suggestion

```

Comment (vgrassia) [reply .github/workflows/build.yml:73]: Unused in `build` workflow.

File: .github/workflows/release.yml [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+      branch-name: ${{ steps.branch.outputs.branch-name }}
+    steps:
+      - name: Branch check
+        run: |
+          if [[ "$GITHUB_REF" != "refs/heads/rc" ]] && [[ "$GITHUB_REF" != "refs/heads/hotfix" ]]; then
+            echo "==================================="
+            echo "[!] Can only release from the 'rc' or 'hotfix' branches"
+            echo "==================================="
+            exit 1
+          fi
```
Comment (vgrassia) [inline .github/workflows/release.yml:23]:
```suggestion
          fi
          
```

File: .github/workflows/release.yml [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+        id: branch
+        run: |
+          BRANCH_NAME=$(basename ${{ github.ref }})
+          echo "::set-output name=branch-name::$BRANCH_NAME"
+
+  release-docker:
+    name: Build Docker images
+    runs-on: ubuntu-20.04
+    needs:
+      - setup
```
Comment (vgrassia) [inline .github/workflows/release.yml:51]:
```suggestion
    needs: setup
```

File: .github/workflows/release.yml [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+
+      - name: Checkout repo
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579
+
+      - name: Setup service name
+        id: setup
+        run: |
+          SERVICE_NAME="key-connector"
+          echo "SERVICE_NAME: $SERVICE_NAME"
+          echo "::set-output name=service_name::$SERVICE_NAME"
```
Comment (vgrassia) [inline .github/workflows/release.yml:78]:
```suggestion

```

Comment (vgrassia) [reply .github/workflows/release.yml:78]: Remove and replace with job `env`.

File: .github/workflows/release.yml [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+          BRANCH_NAME=$(basename ${{ github.ref }})
+          echo "::set-output name=branch-name::$BRANCH_NAME"
+
+  release-docker:
+    name: Build Docker images
+    runs-on: ubuntu-20.04
+    needs:
+      - setup
+    env:
+      _RELEASE_VERSION: ${{ needs.setup.outputs.release_version }}
```
Comment (vgrassia) [inline .github/workflows/release.yml:53]:
```suggestion
      _SERVICE_NAME: key-connector
      _RELEASE_VERSION: ${{ needs.setup.outputs.release_version }}
```

File: .github/workflows/release.yml [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+        id: setup
+        run: |
+          SERVICE_NAME="key-connector"
+          echo "SERVICE_NAME: $SERVICE_NAME"
+          echo "::set-output name=service_name::$SERVICE_NAME"
+
+      - name: Pull latest selfhost image
+        env:
+          SERVICE_NAME: ${{ steps.setup.outputs.service_name }}
+        run: docker pull bitwarden/$SERVICE_NAME:$_BRANCH_NAME
```
Comment (vgrassia) [inline .github/workflows/release.yml:83]:
```suggestion
        run: docker pull bitwarden/$_SERVICE_NAME:$_BRANCH_NAME
```

File: .github/workflows/release.yml [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+      - name: Tag version and latest
+        env:
+          SERVICE_NAME: ${{ steps.setup.outputs.service_name }}
+        run: |
+          docker tag bitwarden/$SERVICE_NAME:$_BRANCH_NAME bitwarden/$SERVICE_NAME:$_RELEASE_VERSION
+          docker tag bitwarden/$SERVICE_NAME:$_BRANCH_NAME bitwarden/$SERVICE_NAME:latest
```
Comment (vgrassia) [inline .github/workflows/release.yml:90]:
```suggestion
        run: |
          docker tag bitwarden/$_SERVICE_NAME:$_BRANCH_NAME bitwarden/$_SERVICE_NAME:$_RELEASE_VERSION
          docker tag bitwarden/$_SERVICE_NAME:$_BRANCH_NAME bitwarden/$_SERVICE_NAME:latest
```

File: .github/workflows/release.yml [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+        run: docker images
+
+      - name: Push version and latest image
+        env:
+          DOCKER_CONTENT_TRUST: 1
+          DOCKER_CONTENT_TRUST_REPOSITORY_PASSPHRASE: ${{ steps.setup-dct.outputs.dct-delegate-repo-passphrase }}
+          SERVICE_NAME: ${{ steps.setup.outputs.service_name }}
+        run: |
+          docker push bitwarden/$SERVICE_NAME:$_RELEASE_VERSION
+          docker push bitwarden/$SERVICE_NAME:latest
```
Comment (vgrassia) [inline .github/workflows/release.yml:102]:
```suggestion
        run: |
          docker push bitwarden/$_SERVICE_NAME:$_RELEASE_VERSION
          docker push bitwarden/$_SERVICE_NAME:latest
```

File: .github/workflows/release.yml [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+
+      - name: Log out of Docker
+        run: docker logout
+
+  release-github:
+    name: Create GitHub Release
+    runs-on: ubuntu-20.04
+    needs:
+      - setup
```
Comment (vgrassia) [inline .github/workflows/release.yml:111]:
```suggestion
    needs: setup
```

File: .github/workflows/release.yml [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+      - name: Create release
+        uses: ncipollo/release-action@40bb172bd05f266cf9ba4ff965cb61e9ee5f6d01
+        with:
+          commit: ${{ github.sha }}
+          tag: "v${{ needs.setup.outputs.release_version }}"
+          name: "Version ${{ needs.setup.outputs.release_version }}"
+          body: "<insert release notes here>"
+          token: ${{ secrets.GITHUB_TOKEN }}
+          draft: true
+
```
Comment (vgrassia) [inline .github/workflows/release.yml:122]:
```suggestion


```

Comment (vgrassia) [reply .github/workflows/release.yml:122]: Two blank lines between jobs.

File: .github/workflows/release.yml [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+          DOCKER_CONTENT_TRUST: 1
+          DOCKER_CONTENT_TRUST_REPOSITORY_PASSPHRASE: ${{ steps.setup-dct.outputs.dct-delegate-repo-passphrase }}
+          SERVICE_NAME: ${{ steps.setup.outputs.service_name }}
+        run: |
+          docker push bitwarden/$SERVICE_NAME:$_RELEASE_VERSION
+          docker push bitwarden/$SERVICE_NAME:latest
+
+      - name: Log out of Docker
+        run: docker logout
+
```
Comment (vgrassia) [inline .github/workflows/release.yml:106]:
```suggestion


```

Comment (vgrassia) [reply .github/workflows/release.yml:106]: Two blank lines between jobs.

File: .github/workflows/release.yml [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+    name: Check for failures
+    if: always()
+    runs-on: ubuntu-20.04
+    needs:
+      - release-docker
+      - release-github
+      - setup
+    steps:
+      - name: Check if any job failed
+        if: ${{ (github.ref == 'refs/heads/master') || (github.ref == 'refs/heads/rc') }}
```
Comment (vgrassia) [inline .github/workflows/release.yml:133]:
```suggestion
        if: ${{ (github.ref == 'refs/heads/master') || (github.ref == 'refs/heads/rc') || (github.ref == 'refs/heads/hotfix') }}
```

**Review by vgrassia — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml (hunk 1 above)
Comment (mimartin12) [reply .github/workflows/build.yml]: Now that I am using the same linter as you are, I was able to identify these easier. I added the correct formatting as well as appended versions to actions that had them.

File: .github/workflows/release.yml [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+    steps:
+      - name: Branch check
+        run: |
+          if [[ "$GITHUB_REF" != "refs/heads/rc" ]] && [[ "$GITHUB_REF" != "refs/heads/hotfix" ]]; then
+            echo "==================================="
+            echo "[!] Can only release from the 'rc' or 'hotfix' branches"
+            echo "==================================="
+            exit 1
+          fi
+
```
Comment (vgrassia) [inline .github/workflows/release.yml:24]:
```suggestion

```

Comment (vgrassia) [reply .github/workflows/release.yml:24]: Remove trailing spaces.

File: .github/workflows/release.yml [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+    name: Check for failures
+    if: always()
+    runs-on: ubuntu-20.04
+    needs:
+      - release-docker
+      - release-github
+      - setup
+    steps:
+      - name: Check if any job failed
+        if: ${{ (github.ref == 'refs/heads/master') || (github.ref == 'refs/heads/rc') || (github.ref == 'refs/heads/hotfix') }}
```
Comment (vgrassia) [inline .github/workflows/release.yml:124]:
```suggestion
        if: |
          github.ref == 'refs/heads/master'
          || github.ref == 'refs/heads/rc'
          || github.ref == 'refs/heads/hotfix'
```

Comment (vgrassia) [reply .github/workflows/release.yml:124]: Line length over 120.

**Review by vgrassia — CHANGES_REQUESTED** [review]

**Review by vgrassia — APPROVED** [review]

**Pull request merged**